### DISCUSSION TITLE
[lodash] /fp get/prop return type always contains undefined, even when it would be otherwise impossible

### DISCUSSION THREAD

**Discussion by yangm97** [DefinitelyTyped/DefinitelyTyped#67435] [Issues with a @types package]
Take for instance this redux selector:

```ts
const selectChannelType = createCachedSelector(
  selectChannelById,
  c => c.type,
)((s, {id}) => id);
```

The return type is never undefined or null. However, if I try to save some keystrokes with `get` from `lodash/fp`:

```ts
const selectChannelType = createCachedSelector(
  selectChannelById,
  get('type'),
)((s, {id}) => id);
```

The return type now consists of `ChannelType | undefined`, which is technically impossible.

I also tried to force the proper type using generics, which would be longer/uglier but serves as a sanity check, however `undefined` still creeps in the resulting type.

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
Without knowing how `createCachedSelector` is defined, it's impossible to know what's going on here. But when you call `get('type')`, lodash doesn't know any of the type information of your object. You could use explicit generic params to give it the type information, although it sounds like that didn't help either.

In this case I don't think it's feasible for `get()` to know the correct types, so I'd recommend using `c => c.type` instead (which should also have better performance). These days there is little reason to use `get()` since you can use the [optional chaining operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) (`?.`) to handle null/undefined values instead.
